akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpMessageParser.scala

```diff
@@ -42,19 +43,33 @@ private[http] abstract class HttpMessageParser[Output >: MessageOutput <: Parser
 
   def isTerminated = terminated
 
-  val stage: PushPullStage[SessionBytes, Output] =
-    new PushPullStage[SessionBytes, Output] {
-      def onPush(input: SessionBytes, ctx: Context[Output]) = handleParserOutput(self.parseSessionBytes(input), ctx)
-      def onPull(ctx: Context[Output]) = handleParserOutput(self.onPull(), ctx)
-      private def handleParserOutput(output: Output, ctx: Context[Output]): SyncDirective =
-        output match {
-          case StreamEnd    ⇒ ctx.finish()
-          case NeedMoreData ⇒ ctx.pull()
-          case x            ⇒ ctx.push(x)
+  val stage = new GraphStage[FlowShape[SessionBytes, Output]] {
+    val in: Inlet[SessionBytes] = Inlet("Parser.in")
+    val out: Outlet[Output] = Outlet("Parser.out")
+    override val shape: FlowShape[SessionBytes, Output] = FlowShape(in, out)
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new GraphStageLogic(shape) with InHandler with OutHandler {
+        override def onPush(): Unit = handleParserOutput(self.parseSessionBytes(grab(in)))
+        override def onPull(): Unit = handleParserOutput(self.onPull())
+
+        override def onUpstreamFinish(): Unit =
+          if (self.onUpstreamFinish()) completeStage()
+          else if (isAvailable(out)) handleParserOutput(self.onPull())
+
+        private def handleParserOutput(output: Output): Unit = {
+          output match {
+            case StreamEnd    ⇒ completeStage()
+            case NeedMoreData ⇒ pull(in)
+            case x            ⇒ push(out, x)
+          }
+          if (isTerminated) completeStage()
         }
-      override def onUpstreamFinish(ctx: Context[Output]): TerminationDirective =
-        if (self.onUpstreamFinish()) ctx.finish() else ctx.absorbTermination()
-    }
+
+        setHandler(in, this)
+        setHandler(out, this)
```

Comment:
Well done (with making the handler `this`), one small improvement here is that it could be `setHandlers(i, out, this)`